besca/_helper.py

```diff
@@ -190,3 +193,98 @@ def get_means(adata,mycat):
         average_obs=None
         fraction_obs=None
     return(average_obs, fraction_obs)
+
+def concate_adata(adata1, adata2):
+    """ concatenate two adata objects based on the observations
+
+    this function also merges the objects saved in .raw and generates a new combined.raw.
+    The obs from adata1 are preserved. Those from adata2 are lost.
+
+    parameters
+    ----------
+    adata1: ´AnnData´
+        first Anndata object that is to be concatenated (the obs and obs_names will be taken from here)
+    adata2: ´AnnData´
+        second Anndata objec that is to be concatenated
+
+    returns
+    -------
+    AnnData
+        returns the AnnData object contained in .raw with all relevant annotation
+
+    """
+
+    #convert both to sparse matrices
+    adata1_X = sparse.csr_matrix(adata1.X)
+    adata2_X = sparse.csr_matrix(adata2.X)
+
+    #hstack (i.e. horizontally stack them)
+    X = sparse.hstack([adata1_X, adata2_X])
+
+    #convert back to sparse matrix (for some reason hstack returns coo matrix)
+    X = sparse.csr_matrix(X)
+
+    #observations remain unchanged so will use obs from adata1
+    if sum(adata1.obs.CELL != adata2.obs.CELL) != 0:
+        sys.exit('The observations in adata1 and adata2 are not the same. These two anndata objects can not be concatenated.')
+
+    obs = adata1.obs
+    obs_names = adata1.obs_names
+
+    #vars, var_names are concatenated from the two dataframes
+    var = concat([adata1.var, adata2.var])
+    var_names = adata1.var_names.tolist() + adata2.var_names.tolist()
+
+    if var.index.tolist() != var_names:
+        sys.exit('var and var_names index does not match. Something went wrong.')
+    
+    #create our complete combined object
+    adata_combined = AnnData(X = X, obs = obs, var = var)
+
+    if adata1.raw is not None:
+
```

Comment:
Hi,
I might have missed it but what if adata1.raw is None and not adata2.raw ? 
Proposition 
if  any([adata1.raw is not None, adata2.raw is not None]) :
    if any([adata1.raw  is None, adata2.raw is None]):
        sys.exit( ''Only one of the anndata objects contains a .raw!')

Reply:
Agree.

Reply:
fixed in 613d4dd

Reply:
added just a small commit as there was a typo in the fix i proposed... see ad3477f